File: csharp/test/Test/ProviderClient.cs
using Nethereum.Util;

namespace Kios.Service.Sdk.Test;

using Client;
using Types;
using Utils;
using System.Numerics;

public class ProviderClientTests
{
    private NetWorkType network = NetWorkType.ACC_TestNet;
    private ProviderClient providerClient;
    private ProviderClient agentClient;

    public ProviderClientTests()
    {
        providerClient =
            new ProviderClient(network, "0x70438bc3ed02b5e4b76d496625cb7c06d6b7bf4362295b16fdfe91a046d4586c");
        agentClient = new ProviderClient(network, "0x44868157d6d3524beb64c6ae41ee6c879d03c19a357dadb038fefea30e23cbab");
    }

    [SetUp]
    public void Setup()
    {
    }

    [Test]
    public async Task Test01_IsProvider()
    {
        var value = await providerClient.IsProvider(providerClient.Address);
        Assert.That(value, Is.EqualTo(true));
    }

    [Test]
    public async Task Test02_ClearAgent()
    {
        await providerClient.SetAgent(AddressUtil.ZERO_ADDRESS);
        Assert.That(await providerClient.GetAgent(), Is.EqualTo(AddressUtil.ZERO_ADDRESS));
    }

    [Test]
    public async Task Test03_ProvideToAddress()
    {
        var receiver = "0xB6f69F0e9e70034ba0578C542476cC13eF739269";
        var res1 = await providerClient.GetBalanceAccount(receiver);
        var oldBalance = res1.Point.Balance;

        var amount = Amount.Make("100").Value;
        await providerClient.ProvideToAddress(providerClient.Address, receiver, amount);
        var res2 = await providerClient.GetBalanceAccount(receiver);

        Assert.That(res2.Point.Balance, Is.EqualTo(BigInteger.Add(oldBalance, amount)));
    }

    [Test]
    public async Task Test04_ProvideToPhone()
    {
        var phoneNumber = "+82 10-9000-5000";
        var res1 = await providerClient.GetBalancePhone(phoneNumber);
        var oldBalance = res1.Point.Balance;
        var amount = Amount.Make("100").Value;

        await providerClient.ProvideToPhone(providerClient.Address, phoneNumber, amount);
        var res2 = await providerClient.GetBalancePhone(phoneNumber);

        Assert.That(res2.Point.Balance, Is.EqualTo(BigInteger.Add(oldBalance, amount)));
    }

    [Test]
    public async Task Test05_SetNewAgent()
    {
        await providerClient.SetAgent(agentClient.Address);
        Assert.That(await providerClient.GetAgent(), Is.EqualTo(agentClient.Address));
    }

    [Test]
    public async Task Test06_ProvideToAddress()
    {
        var receiver = "0xB6f69F0e9e70034ba0578C542476cC13eF739269";
        var res1 = await providerClient.GetBalanceAccount(receiver);
        var oldBalance = res1.Point.Balance;

        var amount = Amount.Make("100").Value;
        await agentClient.ProvideToAddress(providerClient.Address, receiver, amount);
        var res2 = await providerClient.GetBalanceAccount(receiver);

        Assert.That(res2.Point.Balance, Is.EqualTo(BigInteger.Add(oldBalance, amount)));
    }

    [Test]
    public async Task Test07_ProvideToPhone()
    {
        var phoneNumber = "+82 10-9000-5000";
        var res1 = await providerClient.GetBalancePhone(phoneNumber);
        var oldBalance = res1.Point.Balance;
        var amount = Amount.Make("100").Value;

        await agentClient.ProvideToPhone(providerClient.Address, phoneNumber, amount);
        var res2 = await providerClient.GetBalancePhone(phoneNumber);

        Assert.That(res2.Point.Balance, Is.EqualTo(BigInteger.Add(oldBalance, amount)));
    }
}

File: java/ver21/sdk/src/test/java/org/kios/service/sdk/client/ProviderClient.java
package org.kios.service.sdk.client;

import org.kios.service.sdk.data.NetWorkType;
import org.kios.service.sdk.data.UserBalance;
import org.kios.service.sdk.utils.Amount;
import org.junit.jupiter.api.Test;
import org.web3j.abi.datatypes.Address;

import java.math.BigInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ProviderClientTest {
    ProviderClient providerClient = new ProviderClient(NetWorkType.acc_testnet, "0x70438bc3ed02b5e4b76d496625cb7c06d6b7bf4362295b16fdfe91a046d4586c");
    ProviderClient agentClient = new ProviderClient(NetWorkType.acc_testnet, "0x44868157d6d3524beb64c6ae41ee6c879d03c19a357dadb038fefea30e23cbab");

    @Test
    void ProviderClient() {
        try {
            // isProvider
            {
                System.out.println("[ isProvider ]");
                boolean value = providerClient.isProvider(providerClient.getAddress());
                assertTrue(value);
            }

            // getBalanceAccount
            {
                System.out.println("[ getBalanceAccount ]");
                UserBalance value = providerClient.getBalanceAccount(providerClient.getAddress());
                assertTrue(value.point.balance.compareTo(BigInteger.ZERO) > 0);
                assertTrue(value.token.balance.compareTo(BigInteger.ZERO) > 0);
            }

            // getBalancePhone
            {
                System.out.println("[ getBalancePhone ]");
                UserBalance value = providerClient.getBalancePhone("+82 10-1000-2000");
                assertTrue(value.point.balance.compareTo(BigInteger.ZERO) > 0);
            }

            // ClearAgent
            {
                System.out.println("[ ClearAgent ]");
                providerClient.setAgent(Address.DEFAULT.toString());
                assertEquals(providerClient.getAgent(), Address.DEFAULT.toString());
            }

            // ProviderToAddress
            {
                System.out.println("[ ProviderToAddress ]");
                Boolean isProvider = providerClient.isProvider(providerClient.getAddress());
                if (isProvider) {
                    String receiver = "0xB6f69F0e9e70034ba0578C542476cC13eF739269";
                    UserBalance res1 = providerClient.getBalanceAccount(receiver);
                    BigInteger oldBalance = res1.point.balance;

                    BigInteger amount = Amount.make("100").getValue();
                    providerClient.provideToAddress(providerClient.getAddress(), receiver, amount);
                    UserBalance res2 = providerClient.getBalanceAccount(receiver);

                    assertEquals(res2.point.balance, oldBalance.add(amount));
                }
            }

            // ProviderToPhone
            {
                System.out.println("[ ProviderToPhone ]");
                Boolean isProvider = providerClient.isProvider(providerClient.getAddress());
                if (isProvider) {
                    String phoneNumber = "+82 10-9000-5000";
                    UserBalance res1 = providerClient.getBalancePhone(phoneNumber);
                    BigInteger oldBalance = res1.point.balance;

                    BigInteger amount = Amount.make("100").getValue();
                    providerClient.provideToPhone(providerClient.getAddress(), phoneNumber, amount);
                    UserBalance res2 = providerClient.getBalancePhone(phoneNumber);

                    assertEquals(res2.point.balance, oldBalance.add(amount));
                }
            }

            // SetAgent
            {
                System.out.println("[ SetAgent ]");
                providerClient.setAgent(agentClient.getAddress());
                assertEquals(providerClient.getAgent().toLowerCase(), agentClient.getAddress().toLowerCase());
            }


            // ProviderToAddressByAgent
            {
                System.out.println("[ ProviderToAddressByAgent ]");
                Boolean isProvider = providerClient.isProvider(providerClient.getAddress());
                if (isProvider) {
                    String receiver = "0xB6f69F0e9e70034ba0578C542476cC13eF739269";
                    UserBalance res1 = providerClient.getBalanceAccount(receiver);
                    BigInteger oldBalance = res1.point.balance;

                    BigInteger amount = Amount.make("100").getValue();
                    agentClient.provideToAddress(providerClient.getAddress(), receiver, amount);
                    UserBalance res2 = providerClient.getBalanceAccount(receiver);

                    assertEquals(res2.point.balance, oldBalance.add(amount));
                }
            }

            // ProviderToPhoneByAgent
            {
                System.out.println("[ ProviderToPhoneByAgent ]");
                Boolean isProvider = providerClient.isProvider(providerClient.getAddress());
                if (isProvider) {
                    String phoneNumber = "+82 10-9000-5000";
                    UserBalance res1 = providerClient.getBalancePhone(phoneNumber);
                    BigInteger oldBalance = res1.point.balance;

                    BigInteger amount = Amount.make("100").getValue();
                    agentClient.provideToPhone(providerClient.getAddress(), phoneNumber, amount);
                    UserBalance res2 = providerClient.getBalancePhone(phoneNumber);

                    assertEquals(res2.point.balance, oldBalance.add(amount));
                }
            }
        } catch (Exception e) {
            assertEquals("some exception message...", e.getMessage());
        }
    }

}
